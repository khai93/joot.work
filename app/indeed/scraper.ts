// Copyright 2022 Nang Khai.  All rights reserved.
// Use of this source code is governed by a GNU General Public License v3.0
// license that can be found in the LICENSE file.

import { JobPost } from "@/core/JobPostService";
import { JobSearchService, JobSearchFilter } from "@/core/JobSearchService";
import { injectable } from "inversify";
import { parseFetchResponseHTML, parseFetchResponseText } from "../util/parseFetchResponse";

export type IndeedParseResult = {
    companyBrandingAttributes?: {
        logoUrl?: string,
        headerImageUrl?: string
    },
    company: string,
    companyRating: number,
    companyOverviewLink: string,
    expired: boolean,
    title: string,
    urgentlyHiring: boolean,
    viewJobLink: string,
    remoteLocation: boolean,
    formattedRelativeTime: string
}

@injectable()
export class IndeedJobScraper implements JobSearchService {
    async search(keywords: string[], filters?: JobSearchFilter): Promise<JobPost[]> {
        const requestURL = new URL("https://www.indeed.com/jobs");
        requestURL.searchParams.append("q", keywords.join(" "));
        requestURL.searchParams.append("start", filters?.page ? String(Math.max((filters.page - 1) * 10, 0)) : '0');
        
        if (filters?.remoteType != undefined) {
            if (filters.remoteType === 1) {
                requestURL.searchParams.append("sc", "0kf%3Aattr(DSQF7)%3B");
            }
            if (filters.remoteType === 2) {
                requestURL.searchParams.append("sc", "0kf%3Aattr(VAMUB)%3B");
            } 
        }

        if (filters?.location != null && filters?.location !== "undefined") {
            requestURL.searchParams.append("l", filters!.location!);
        }

        const { _body } = await parseFetchResponseText(await fetch(requestURL.toString(), { 
            headers: {
                "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:103.0) Gecko/20100101 Firefox/103.0",
                "Host": requestURL.host,
                "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8"
            },
            credentials: "include"
        }));

        const output: JobPost[] = [];

        const mainMatch = _body.match(/window\.mosaic\.providerData\[\"mosaic-provider-jobcards\"]=\{(.*)\};/m);

        if (mainMatch == null) {
            return [];
        }

        for (let result of Array.from(JSON.parse(`{${mainMatch[1]}}`).metaData.mosaicProviderJobCardsModel.results) as IndeedParseResult[]) {
            result = result as any;
            output.push({
                job_title: result.title,
                company: {
                    company_name: result.company,
                    company_link: new URL("https://www.indeed.com" + result.companyOverviewLink),
                    logoUrl: new URL(result.companyBrandingAttributes?.logoUrl || "http://placehold.jp/64x64.png")
                },
                job_link: new URL("https://www.indeed.com" + result.viewJobLink),
                postedDate: null,
                formattedDate: result.formattedRelativeTime
            });
        }

        return output
    }
}