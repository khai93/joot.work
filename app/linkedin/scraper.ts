// Copyright 2022 Nang Khai.  All rights reserved.
// Use of this source code is governed by a GNU General Public License v3.0
// license that can be found in the LICENSE file.

import { JobPost } from "@/core/JobPostService";
import { JobScraperService, JobSearchFilter } from "@/core/JobScraperService";
import { injectable } from "inversify";
import { parseFetchResponseHTML } from "../util/parseFetchResponse";

@injectable()
export class LinkedinJobScraper implements JobScraperService {
    constructor() {}

    async search(keywords: string[], filters?: JobSearchFilter): Promise<JobPost[]> {
        const requestURL = new URL('https://www.linkedin.com/jobs/search/');
        requestURL.searchParams.append("keywords", keywords.join(" "));

        const { _document } = await parseFetchResponseHTML(await fetch(requestURL.toString(), { 
            headers: {
                "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/42.0.2311.135 Safari/537.36 Edge/12.246",
                "Host": requestURL.host
            },
            credentials: "same-origin"
        }));

        const output: JobPost[] = [];
        
        const searchResults = _document.getElementsByClassName("jobs-search__results-list")[0].children;
        
        for (let i = 0;i < searchResults.length;i++) {
            const subtitle = searchResults[i].querySelector(".base-search-card__subtitle");
            if (subtitle?.firstElementChild == null) continue;
            output.push({
                job_title: searchResults[i]!.querySelector(".base-search-card__title")!.textContent!.trim(),
                company: {
                    company_name: subtitle!.firstElementChild!.innerHTML.trim(),
                    company_link: new URL(subtitle!.firstElementChild!.getAttribute("href")!),
                    logoUrl: new URL(searchResults[i]!.querySelector(".artdeco-entity-image")!.getAttribute("data-delayed-url")!)
                },
                postedDate: new Date(searchResults[i]!.querySelector("time")!.getAttribute("datetime")!),
                formattedDate: null,
                job_link: new URL(searchResults[i]!.querySelector(".base-card__full-link")!.getAttribute("href")!)
            });
        }

        return output;
    }
}