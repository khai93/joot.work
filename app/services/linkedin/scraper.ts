// Copyright 2022 Nang Khai.  All rights reserved.
// Use of this source code is governed by a GNU General Public License v3.0
// license that can be found in the LICENSE file.

import { JobPost } from "@/core/JobPost";
import { JobScraperService, JobSearchFilter } from "../../core/JobScraperService";
import { injectable } from "inversify";

import { AxiosInstance } from "axios";

@injectable()
export class LinkedinJobScraper implements JobScraperService {
    constructor(
    ) {}

    search(keywords: string[], filters?: JobSearchFilter): Promise<JobPost[]> {
        throw new Error("Method not implemented.");
    }
}