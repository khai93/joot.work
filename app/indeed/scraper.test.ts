// Copyright 2022 Nang Khai.  All rights reserved.
// Use of this source code is governed by a GNU General Public License v3.0
// license that can be found in the LICENSE file.

import { rest } from "msw";
import { setupServer } from "msw/lib/node";
import { IndeedJobScraper } from "./scraper";
import fs from "fs/promises";
import path from "path";

const server = setupServer(
    rest.get("https://www.indeed.com/jobs", async (req, res, ctx) => {
        const mockHTMLPage = await fs.readFile(path.resolve(__dirname, "./__mocks__/indeedSearchMock.html"));
        
        return res(ctx.set('Content-Type', 'text/html'), ctx.body(mockHTMLPage));
    })
);

beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());

describe('LinkedinJobScraper', () => { 
    describe('#search', () => {
        it('should return correct test data', async () => {
            const scraper = new IndeedJobScraper();
            expect(await scraper.search(['test'])).toMatchSnapshot();
        });
    })
});