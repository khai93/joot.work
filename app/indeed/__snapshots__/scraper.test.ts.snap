// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LinkedinJobScraper #search should return correct test data 1`] = `
Array [
  Object {
    "company": Object {
      "company_link": "https://www.indeed.com/cmp/Oak-Hall-Cap-&-Gown",
      "company_name": "Oak Hall Cap & Gown",
      "logoUrl": "http://placehold.jp/64x64.png",
    },
    "formattedDate": "1 day ago",
    "job_link": "https://www.indeed.com/m/basecamp/viewjob?viewtype=embedded&jk=a37e986f201b7f55&from=vjs&tk=1g4nf337epv1h800&advn=8579125579062878&adid=391329288&ad=-6NYlbfkN0Bh40mwGfYgYZwXpcEHv5iut404hKWzEBzy-cTkBlb-dx3GDwtdwihb0GsTsT4s3krtMPoHptpJb3J27vK5pUvDi0a649_HB7n9dea9_hQRnGcSIpYp_flfJbEi6Y56zeTkbSGjwnDzBc-ftm-03xBPmiWFoNP7nq5Odm-sIBye8njsuxlungnzWoYQQNebHsRPjMzf2GncqHAkpc-1lLF2U2a5VgEECqaNkTEBz1Yf-gYHdKNnSxwhAC2m1izs9Q_4AnwmpAPq3IrajZ94rSInOGek0-Jx2eIg0nuG1NtX7u26k89_y2EBG9IpCs1UVkKxRv3GioCef77SJr33aSWU3-Jyfb4wEMkoCE4Zq68eWFXT3XTPl_LkhOZnks9VtPkFtCFe3KZHkE1QLXa4lFBpjE6NFv66Scg2OIoc8S4SLm_auguLXf1DlAgaozsiI173jRJNJQds1tjxOfOoPIx2vcmXId-5W9M%3D",
    "job_title": "Entry Level Web Developer",
    "postedDate": null,
  },
  Object {
    "company": Object {
      "company_link": "https://www.indeed.com/cmp/Apple",
      "company_name": "Apple",
      "logoUrl": "https://d2q79iu7y748jz.cloudfront.net/s/_squarelogo/256x256/9d957ded6f74ee20d805109548fb36e8",
    },
    "formattedDate": "Today",
    "job_link": "https://www.indeed.com/m/basecamp/viewjob?viewtype=embedded&jk=19cee7da4693369c&from=vjs&tk=1g4nf337epv1h800&advn=4317892059899432&adid=376525686&ad=-6NYlbfkN0BvKrLyj5gPmtZO9T8euul8TCxuuKNOtzRJOomxnwSEodTz2Bc-sPZlADHp0xxmf8WtgwAMp1M4Ysvhsr1LqTslbraQiaMk1Xtvdy-eoP1oKfrYhU2ZdKatwvfBPVaLnEksndVuZj_-hs22DQzEj01tZbUN7slYnmHsEBqCkpkrht0JtVC9nOANdWBkNxUAin2jCx7F8q8jhxirniM_WFxAg1ZxKEiwj10QT9OOJ1dYsMkh78_Iz6wJYAZEJhApLS-TSWe7D-w5QmUY9VxxE3hH_Z4bRg4EUt9mB7LGAL3q9v8fXnRoOeEM15RqW0S8D1T4-iiwz4VxW0ljr6mXhCqZAa4RVpY66JFe-e8d_nQp2KYLWyDjVZiWSx5m732PwVxviAtCZV_ewNzw1SjFXmHYmut8mx9kZL3Zep6rQUtI6riRaJDv79yeTg5s-PFaA4IUfXqzd0OmM8lPVublNhSuznyYZwIS9i1OuG9UG22kQ44cJHEW-lf6gTmmNjC1jM1-LeUkeEsVD7lqy6RbZBJMQzeNYDVjpPLUzna5cU7LIN0ZuzKGlZfE46p1hB1sVAyJyJvHJtVxT3BgHOcb05paXfHgpnu-IhWQg8KVFrl61URGwumuumxUldDKXerILlav-8-26y7qOzIKbBuptceRAUEKy0810DnMXDyW0IUej-AZwfuuYsCiM3mhXuND4z3cM4izVsimTdD1PhOIuW0Kjz3drC92G5ZVQ1ZO46LecVahhpL-YBl4qOgQT7pmHlXUn_cTzD8yzADB0isyuUrPHHWzFDYCv85e6X0y6--MuLMUnPKzLHMCjk-YIKbLjahGEFIiG9jX86sDLRLl8y0Ux0CS304s5RaAQGXDo7ZeC_iEEMXUh8gMAYr56gelH0QrqAAR68YDBNe4dYsfM13Bp01A1FMR3FQkD9itmrMk_XulWCiGC8Bf",
    "job_title": "Web UI Engineer (Early Career)",
    "postedDate": null,
  },
  Object {
    "company": Object {
      "company_link": "https://www.indeed.comundefined/",
      "company_name": "Digital Gravy",
      "logoUrl": "http://placehold.jp/64x64.png",
    },
    "formattedDate": "30+ days ago",
    "job_link": "https://www.indeed.com/m/basecamp/viewjob?viewtype=embedded&jk=03c04e321747f745&from=vjs&tk=1g4nf337epv1h800",
    "job_title": "Front End Web Developer (Oxygen Builder - Training Provided)",
    "postedDate": null,
  },
  Object {
    "company": Object {
      "company_link": "https://www.indeed.com/cmp/Indeed",
      "company_name": "Indeed",
      "logoUrl": "https://d2q79iu7y748jz.cloudfront.net/s/_squarelogo/256x256/b22fcd0b10d4304b4180de07be1d5f83",
    },
    "formattedDate": "30+ days ago",
    "job_link": "https://www.indeed.com/m/basecamp/viewjob?viewtype=embedded&jk=f0cafa676143e05f&from=vjs&tk=1g4nf337epv1h800&advn=9767451333976163&adid=383080361&ad=-6NYlbfkN0CiRNM7CVr8YueLFKlzwbFWI0o7IjV438l4sVrvKZ0flpURU_mqoI8E88RAJZx1_nSgG-OOhldI3ZbVZdOMxU4RaEDaYoCj6Q_I5xqgjVrY9hFZyt5con6aqydZ-L3mTVxvh0Oqt8gnP6nznnrL0JUZoOPydVx7bkB8AUXu61ggF5myMzxWq63o2aZF82vXOcRIe6UBtI-WKEb5kBz_LPEPu76pMkrvW4aWPIg-hMSufhk1uKXm1gesVzZUBdJqRoYVHS0-QAoAUY1TRE_le9YDN04_ErMkMBOZhiDwD-rqSMHnuMM2qjU9vhs7Y8TXhoFOpz6-tfgunlYTTbzzLuNOP3ykU-KWRsE59sFcHP_CmAuc252BinLNUOn8CswDuYapfS5A54LylV-M7He0730Luy-IaQYRYphpk23tM06YvidO9fjLu-O9H1PLVvfw2_6dP81YDuhX4wvk0n8w2pkOzoPrpGqDeJH_-kVdiyJb_TrLbOMVixLbjofINNhQPuUCmTAgms0tSoIQBNaZnrC8",
    "job_title": "Web Developer",
    "postedDate": null,
  },
  Object {
    "company": Object {
      "company_link": "https://www.indeed.comundefined/",
      "company_name": "Booli Project",
      "logoUrl": "http://placehold.jp/64x64.png",
    },
    "formattedDate": "23 days ago",
    "job_link": "https://www.indeed.com/m/basecamp/viewjob?viewtype=embedded&jk=e314806960b703f4&from=vjs&tk=1g4nf337epv1h800",
    "job_title": "React Web Developer",
    "postedDate": null,
  },
  Object {
    "company": Object {
      "company_link": "https://www.indeed.com/cmp/Upshift",
      "company_name": "Upshift",
      "logoUrl": "http://placehold.jp/64x64.png",
    },
    "formattedDate": "2 days ago",
    "job_link": "https://www.indeed.com/m/basecamp/viewjob?viewtype=embedded&jk=4e11beb75d93d8c2&from=vjs&tk=1g4nf337epv1h800",
    "job_title": "Frontend Developer (React.js)",
    "postedDate": null,
  },
  Object {
    "company": Object {
      "company_link": "https://www.indeed.com/cmp/Amazon.com",
      "company_name": "Amazon.com Services LLC",
      "logoUrl": "http://placehold.jp/64x64.png",
    },
    "formattedDate": "30+ days ago",
    "job_link": "https://www.indeed.com/m/basecamp/viewjob?viewtype=embedded&jk=1cc5a9d58d452365&from=vjs&tk=1g4nf337epv1h800",
    "job_title": "Reduced-Hours Front-End Engineer",
    "postedDate": null,
  },
  Object {
    "company": Object {
      "company_link": "https://www.indeed.com/cmp/Nintendo",
      "company_name": "Nintendo of America Inc.",
      "logoUrl": "http://placehold.jp/64x64.png",
    },
    "formattedDate": "22 days ago",
    "job_link": "https://www.indeed.com/m/basecamp/viewjob?viewtype=embedded&jk=764c4878abb7962d&from=vjs&tk=1g4nf337epv1h800",
    "job_title": "Software Engineer I (NST)",
    "postedDate": null,
  },
  Object {
    "company": Object {
      "company_link": "https://www.indeed.comundefined/",
      "company_name": "Velocity Promotions LLC",
      "logoUrl": "http://placehold.jp/64x64.png",
    },
    "formattedDate": "3 days ago",
    "job_link": "https://www.indeed.com/m/basecamp/viewjob?viewtype=embedded&jk=e29a48bee6b030f6&from=vjs&tk=1g4nf337epv1h800",
    "job_title": "Junior Full Stack Web Developer - Remote Work Available",
    "postedDate": null,
  },
  Object {
    "company": Object {
      "company_link": "https://www.indeed.comundefined/",
      "company_name": "Salad Technologies",
      "logoUrl": "http://placehold.jp/64x64.png",
    },
    "formattedDate": "11 days ago",
    "job_link": "https://www.indeed.com/m/basecamp/viewjob?viewtype=embedded&jk=4f95106f425fdb49&from=vjs&tk=1g4nf337epv1h800",
    "job_title": "Web Developer",
    "postedDate": null,
  },
  Object {
    "company": Object {
      "company_link": "https://www.indeed.com/cmp/Mutual-of-Omaha",
      "company_name": "Mutual of Omaha",
      "logoUrl": "https://d2q79iu7y748jz.cloudfront.net/s/_squarelogo/256x256/2f69ea819297b4628fa6a0a234854e18",
    },
    "formattedDate": "2 days ago",
    "job_link": "https://www.indeed.com/m/basecamp/viewjob?viewtype=embedded&jk=0adea9d26d0fb2de&from=vjs&tk=1g4nf337epv1h800",
    "job_title": "Frontend Developer (React)",
    "postedDate": null,
  },
  Object {
    "company": Object {
      "company_link": "https://www.indeed.com/cmp/Aston-Technologies-Inc.",
      "company_name": "Aston Technologies Inc.",
      "logoUrl": "http://placehold.jp/64x64.png",
    },
    "formattedDate": "30+ days ago",
    "job_link": "https://www.indeed.com/m/basecamp/viewjob?viewtype=embedded&jk=76785349a91f064f&from=vjs&tk=1g4nf337epv1h800",
    "job_title": "Full Stack Web Developer - Entry Level",
    "postedDate": null,
  },
  Object {
    "company": Object {
      "company_link": "https://www.indeed.comundefined/",
      "company_name": "Observable",
      "logoUrl": "http://placehold.jp/64x64.png",
    },
    "formattedDate": "16 days ago",
    "job_link": "https://www.indeed.com/m/basecamp/viewjob?viewtype=embedded&jk=9a0a93ef4467364d&from=vjs&tk=1g4nf337epv1h800",
    "job_title": "Senior Full Stack Web Developer - Fully Remote, Observablehq.com",
    "postedDate": null,
  },
  Object {
    "company": Object {
      "company_link": "https://www.indeed.com/cmp/Indeed",
      "company_name": "Indeed",
      "logoUrl": "https://d2q79iu7y748jz.cloudfront.net/s/_squarelogo/256x256/b22fcd0b10d4304b4180de07be1d5f83",
    },
    "formattedDate": "5 days ago",
    "job_link": "https://www.indeed.com/m/basecamp/viewjob?viewtype=embedded&jk=a5fb70571f7d7715&from=vjs&tk=1g4nf337epv1h800",
    "job_title": "Research Engineer I",
    "postedDate": null,
  },
  Object {
    "company": Object {
      "company_link": "https://www.indeed.comundefined/",
      "company_name": "Rex",
      "logoUrl": "http://placehold.jp/64x64.png",
    },
    "formattedDate": "1 day ago",
    "job_link": "https://www.indeed.com/m/basecamp/viewjob?viewtype=embedded&jk=d8d5feeb07f07451&from=vjs&tk=1g4nf337epv1h800",
    "job_title": "Junior Software Engineer",
    "postedDate": null,
  },
]
`;
